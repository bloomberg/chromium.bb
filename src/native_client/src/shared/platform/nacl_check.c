/*
 * Copyright (c) 2012 The Native Client Authors. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

/*
 * NaCl service runtime, support for check macros.
 */

#include "native_client/src/shared/platform/nacl_check.h"

#if defined(_DEBUG)
int nacl_check_debug_mode = 1;
#else
int nacl_check_debug_mode = 0;
#endif

void NaClCheckSetDebugMode(int mode) {
  nacl_check_debug_mode = mode;
}

void NaClCheckIntern(const char *fmt, ...) {
  va_list ap;

  va_start(ap, fmt);
  NaClLogV(LOG_FATAL, fmt, ap);
  va_end(ap);
}
