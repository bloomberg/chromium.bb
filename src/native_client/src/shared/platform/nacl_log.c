/*
 * Copyright (c) 2012 The Native Client Authors. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

/*
 * NaCl Server Runtime logging code.
 */
#include "native_client/src/include/nacl_compiler_annotations.h"
#include "native_client/src/include/portability.h"
#include "native_client/src/include/portability_io.h"
#include "native_client/src/include/portability_process.h"
#include "native_client/src/include/portability_string.h"

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <limits.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <errno.h>

#include "native_client/src/shared/platform/nacl_log.h"
#include "native_client/src/shared/platform/nacl_log_intern.h"

#define THREAD_SAFE_DETAIL_CHECK  0
/*
 * If set, check detail_level without grabbing a mutex.  This makes
 * logging much cheaper, but implies that the verbosity level should
 * only be changed prior to going multithreaded.
 */

#include "native_client/src/shared/gio/gio.h"
#include "native_client/src/shared/platform/nacl_exit.h"
#include "native_client/src/shared/platform/nacl_sync.h"
#include "native_client/src/shared/platform/nacl_sync_checked.h"
#include "native_client/src/shared/platform/nacl_threads.h"
#include "native_client/src/shared/platform/nacl_timestamp.h"

static int              g_initialized = 0;

/*
 * All logging output is protected by this mutex, so that the logging
 * code may output using multiple calls to the gio functions and not
 * have logging output from multiple threads get intermixed.
 */
static struct NaClMutex log_mu;
static int              tag_output = 0;
static int              g_abort_count = 0;
/*
 * g_abort_count is incremented prior to calling
 * gNaClLogAbortBehavior, so that the abort hook can invoke other
 * functions that might need to log, without those log entries
 * deadlocking or polluting the crash log output.
 */
static int              g_abort_behavior_active = 0;

#define NACL_VERBOSITY_UNSET INT_MAX

static int              verbosity = NACL_VERBOSITY_UNSET;
static struct Gio       *log_stream = NULL;
static struct GioFile   log_file_stream;
static int              timestamp_enabled = 1;

/* global, but explicitly not exposed in non-test header file */
void (*gNaClLogAbortBehavior)(void) = NaClAbort;

static FILE *NaClLogFileIoBufferFromFile(char const *log_file) {
  int   log_desc;
  FILE  *log_iob;

  log_desc = open(log_file, O_WRONLY | O_APPEND | O_CREAT, 0777);
  if (-1 == log_desc) {
    perror("NaClLogSetFile");
    fprintf(stderr, "Could not create log file\n");
    NaClAbort();
  }

  log_iob = FDOPEN(log_desc, "a");
  if (NULL == log_iob) {
    perror("NaClLogSetFile");
    fprintf(stderr, "Could not fdopen log stream\n");
    NaClAbort();
  }
  return log_iob;
}

/*
 * Setting the log stream buffering to fully buffered, so that the
 * write of the tag string will be less likely to be separated
 * from the write of the actual log message.
 */
static FILE *NaClLogDupFileIo(FILE *orig) {
  int  d;
  FILE *copy;

  /*
   * On windows (at least on a win7 machine which i tested on),
   * fileno(stderr) is -2.  I/O to the stderr stream appears to
   * succeed -- though who knows, maybe fclose(stderr) would actually
   * report an error? -- but DUP of -2 fails.  We don't try to detect
   * -2 (or other windows magic values) as a special case here, since
   * in the future other FILE* might be used here.  Instead, we just
   * check for DUP failure and trundle on as best as we could.
   */
  if (-1 == (d = DUP(fileno(orig)))) {
    copy = orig;
    /* this means that setvbuf later will affect the shared stream */
  } else if (NULL == (copy = FDOPEN(d, "a"))) {
    copy = orig;
    /* ditto */
  }
  (void) setvbuf(copy, (char *) NULL, _IOFBF, 1024);
  return copy;
}

static struct Gio *NaClLogGioFromFileIoBuffer(FILE *log_iob) {
  struct GioFile *log_gio;

  log_gio = malloc(sizeof *log_gio);
  if (NULL == log_gio) {
    perror("NaClLogSetFile");
    fprintf(stderr, "No memory for log buffers\n");
    NaClAbort();
  }
  if (!GioFileRefCtor(log_gio, log_iob)) {
    fprintf(stderr, "NaClLog module internal error: GioFileRefCtor failed\n");
    NaClAbort();
  }
  return (struct Gio *) log_gio;
}

void NaClLogSetFile(char const *log_file) {
  NaClLogSetGio(NaClLogGioFromFileIoBuffer(
      NaClLogFileIoBufferFromFile(log_file)));
}

int NaClLogDefaultLogVerbosity(void) {
  char *env_verbosity;

  if (NULL != (env_verbosity = getenv("NACLVERBOSITY"))) {
    return strtol(env_verbosity, (char **) 0, 0);
  }
  return 0;
}

struct Gio *NaClLogDefaultLogGio(void) {
  char            *log_file;
  FILE            *log_iob;

  log_file = getenv("NACLLOG");

  if (NULL == log_file) {
    log_iob = NaClLogDupFileIo(stderr);
  } else {
    log_iob = NaClLogFileIoBufferFromFile(log_file);
  }
  return NaClLogGioFromFileIoBuffer(log_iob);
}

void NaClLogModuleInitExtended(int        initial_verbosity,
                               struct Gio *log_gio) {
  if (!g_initialized) {
    NaClXMutexCtor(&log_mu);
    g_initialized = 1;
  }
  NaClLogSetVerbosity(initial_verbosity);
  NaClLogSetGio(log_gio);
}

void NaClLogModuleInit(void) {
  NaClLogModuleInitExtended(NaClLogDefaultLogVerbosity(),
                            NaClLogDefaultLogGio());
}

void NaClLogModuleFini(void) {
  NaClMutexDtor(&log_mu);
  g_initialized = 0;
}

void NaClLogTagNext_mu(void) {
  tag_output = 1;
}

void NaClLogLock(void) {
  NaClXMutexLock(&log_mu);
  NaClLogTagNext_mu();
}

void NaClLogUnlock(void) {
  int run_abort_behavior = 0;
  switch (g_abort_count) {
    case 0:
      NaClXMutexUnlock(&log_mu);
      break;
    case 1:
      /*
       * include an easy-to-recognize output for the fuzzer to recognize
       */
      if (!g_abort_behavior_active) {
        NaClLog_mu(LOG_ERROR, "LOG_FATAL abort exit\n");
        g_abort_behavior_active = 1;
        run_abort_behavior = 1;
        /*
         * run abort behavior only on edge transition when
         * g_abort_behavior_active is first set.
         */
      }
      NaClXMutexUnlock(&log_mu);
      if (run_abort_behavior) {
#ifdef __COVERITY__
        NaClAbort();  /* help coverity figure out that this is the default */
#else
        NaClLogRunAbortBehavior();
#endif
        /* The abort behavior hook may not abort, so abort here in case. */
        NaClAbort();
      }
      break;
    default:
      /*
       * Abort handling code in turn aborted.  Eeep!
       */
      NaClAbort();
      break;
  }
}

static INLINE struct Gio *NaClLogGetGio_mu(void) {
  if (NULL == log_stream) {
    (void) GioFileRefCtor(&log_file_stream, NaClLogDupFileIo(stderr));
    log_stream = (struct Gio *) &log_file_stream;
  }
  return log_stream;
}

static void NaClLogSetVerbosity_mu(int verb) {
  verbosity = verb;
}

void NaClLogPreInitSetVerbosity(int verb) {
  /*
   * The lock used by NaClLogLock has not been initialized and cannot
   * be used; however, prior to initialization we are not going to be
   * invoked from multiple threads, since the caller is responsible
   * for not invoking NaClLog module functions (except for the PreInit
   * ones, obviously) at all, let alone from multiple threads.  Ergo,
   * it is safe to manipulate the module globals without locking.
   */
  NaClLogSetVerbosity_mu(verb);
}

void  NaClLogSetVerbosity(int verb) {
  NaClLogLock();
  NaClLogSetVerbosity_mu(verb);
  NaClLogUnlock();
}

void  NaClLogIncrVerbosity(void) {
  NaClLogLock();
  if (NACL_VERBOSITY_UNSET == verbosity) {
    verbosity = 0;
  }
  ++verbosity;
  NaClLogUnlock();
}

int NaClLogGetVerbosity(void) {
  int v;

  NaClLogLock();
  if (NACL_VERBOSITY_UNSET == verbosity) {
    verbosity = 0;
  }
  v = verbosity;
  NaClLogUnlock();

  return v;
}

static void NaClLogSetGio_mu(struct Gio *stream) {
  if (NULL != log_stream) {
    (void) (*log_stream->vtbl->Flush)(log_stream);
  }
  log_stream = stream;
}

void NaClLogPreInitSetGio(struct Gio *out_stream) {
  /*
   * See thread safety comment in NaClLogPreInitSetVerbosity.
   */
  NaClLogSetGio_mu(out_stream);
}

void NaClLogSetGio(struct Gio *stream) {
  NaClLogLock();
  NaClLogSetGio_mu(stream);
  NaClLogUnlock();
}

struct Gio  *NaClLogGetGio(void) {
  struct Gio  *s;

  NaClLogLock();
  s = NaClLogGetGio_mu();
  NaClLogUnlock();

  return s;
}

void NaClLogEnableTimestamp(void) {
  timestamp_enabled = 1;
}

void NaClLogDisableTimestamp(void) {
  timestamp_enabled = 0;
}

static void NaClLogOutputTag_mu(struct Gio *s) {
  char timestamp[128];
  int  pid;

  if (timestamp_enabled && tag_output) {
    pid = GETPID();
    gprintf(s, "[%d,%u:%s] ",
            pid,
            NaClThreadId(),
            NaClTimeStampString(timestamp, sizeof timestamp));
    tag_output = 0;
  }
}

/*
 * Output a printf-style formatted message if the log verbosity level
 * is set higher than the log output's detail level.  Note that since
 * this is not a macro, log message arguments that have side effects
 * will have their side effects regardless of whether the
 * corresponding log message is printed or not.  This is good from a
 * consistency point of view, but it means that should a logging
 * argument be expensive to compute, the log statement needs to be
 * surrounded by something like
 *
 *  if (detail_level <= NaClLogGetVerbosity()) {
 *    NaClLog(detail_level, "format string", expensive_arg(), ...);
 *  }
 *
 * The log message, if written, is prepended by a microsecond
 * resolution timestamp on linux and a millisecond resolution
 * timestamp on windows.  This means that if the NaCl app can read its
 * own logs, it can distinguish which host OS it is running on.
 */
void NaClLogDoLogV_mu(int         detail_level,
                      char const  *fmt,
                      va_list     ap) {
  struct Gio  *s;

  if (0 == g_abort_count) {
    s = NaClLogGetGio_mu();

    NaClLogOutputTag_mu(s);
    (void) gvprintf(s, fmt, ap);
    (void) (*s->vtbl->Flush)(s);
  } else {
    (void) fprintf(stderr, "POST-ABORT: ");
    (void) vfprintf(stderr, fmt, ap);
    (void) fflush(stderr);
  }

  if (LOG_FATAL == detail_level) {
    ++g_abort_count;
  }
}

void NaClLogV_mu(int        detail_level,
                 char const *fmt,
                 va_list    ap) {
  if (NACL_VERBOSITY_UNSET == verbosity) {
    verbosity = NaClLogDefaultLogVerbosity();
  }

  if (detail_level <= verbosity) {
    NaClLogDoLogV_mu(detail_level, fmt, ap);
  }
}

void NaClLogV(int         detail_level,
              char const  *fmt,
              va_list     ap) {
#if !THREAD_SAFE_DETAIL_CHECK
  if (detail_level > verbosity) {
    return;
  }
#endif
  NaClLogLock();
  NaClLogV_mu(detail_level, fmt, ap);
  NaClLogUnlock();
}

void NaClLog(int         detail_level,
             char const  *fmt,
             ...) {
  va_list ap;

#if !THREAD_SAFE_DETAIL_CHECK
  if (NACL_LIKELY(detail_level > verbosity)) {
    return;
  }
#endif

  NaClLogLock();
  va_start(ap, fmt);
  NaClLogV_mu(detail_level, fmt, ap);
  va_end(ap);
  NaClLogUnlock();
}

void NaClLog_mu(int         detail_level,
                char const  *fmt,
                ...) {
  va_list ap;

#if THREAD_SAFE_DETAIL_CHECK
  if (detail_level > verbosity) {
    return;
  }
#endif

  va_start(ap, fmt);
  NaClLogV_mu(detail_level, fmt, ap);
  va_end(ap);
}

void NaClLogSetAbortBehavior(void (*fn)(void)) {
  NaClXMutexLock(&log_mu);
  gNaClLogAbortBehavior = fn;
  NaClXMutexUnlock(&log_mu);
}

void NaClLogRunAbortBehavior(void) {
  (*gNaClLogAbortBehavior)();
}
