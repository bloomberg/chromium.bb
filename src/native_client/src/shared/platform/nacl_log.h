/*
 * Copyright (c) 2012 The Native Client Authors. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

/*
 * NaCl logging module.
 *
 * This module is used in many parts of Native Client to generate
 * logging output.
 *
 * Example usage:
 *
 *   #include "native_client/src/trusted/platform/nacl_log.h"
 *
 *   NaClLog(3, "Hello %s\n", "world");
 *
 * The 3 is a "detail level" for this particular log message, the
 * "Hello %s\n" is the familiar printf-style format string, and all
 * subsequent arguments are the corresponding arguments.  Type
 * checking for these arguments are done when gcc-compatible compilers
 * are used.
 *
 * When running the program (assuming the main function invoked the
 * module initializer(s) appropriately [see NaClLogModuleInit*
 * below]), you can set the global verbosity level by running the
 * program with the NACLVERBOSITY environment variable, e.g.:
 *
 *   NACLVERBOSITY=3
 *
 * Pre-defined logging detail levels of LOG_INFO, LOG_WARNING,
 * LOG_ERROR, and LOG_FATAL may also be used.  When LOG_FATAL is used,
 * the application aborts after the log message is generated.
 * Messages at these levels cannot be suppressed.
 *
 *
 * The default logging output is standard error.  NB: on Windows, both
 * stdout and stderr are discarded for Windowed applications.
 *
 * All logging output may be redirected to a file using the NACLLOG
 * environment variable, the value of which should be the path to a
 * file.  NOTE: when running in Chrome, the outer sandbox is by
 * default enabled, so the log module will be unable to open the log
 * file.  In order to enable this for testing, use the --no-sandbox
 * flag to Chrome.  (This is not recommended for normal use, since it
 * eliminates a layer of defense.)
 */

#ifndef NATIVE_CLIENT_SRC_TRUSTED_PLATFORM_NACL_LOG_H__
#define NATIVE_CLIENT_SRC_TRUSTED_PLATFORM_NACL_LOG_H__

#include <stdarg.h>

#include "native_client/src/include/build_config.h"
#include "native_client/src/include/nacl_base.h"

#ifdef __native_client__
# define ATTRIBUTE_FORMAT_PRINTF(m, n) __attribute__((format(printf, m, n)))
#else
# include "native_client/src/include/portability.h"
#endif

struct Gio;

EXTERN_C_BEGIN

/*
 * PreInit functions may be used to set default module parameter
 * values before the module initializer is called.  This is needed in
 * some cases, such as by users of NaClNrdModuleInit or
 * NaClAllModuleInit, where a list of module initializer is invoked,
 * and the caller wants to crank up logging to get logging output from
 * functions invoked in the module initializers that occur after
 * NaClLogModuleInit (and prior to NaClNrdModuleInit returning).
 */
void NaClLogPreInitSetVerbosity(int verb);
void NaClLogPreInitSetGio(struct Gio *out_stream);

/*
 * NaClLogModuleInit() is idempotent, provided that all the calls to
 * it originate from a single thread.
 */
void NaClLogModuleInit(void);

void NaClLogModuleInitExtended(int        initial_verbosity,
                               struct Gio *log_gio);

/*
 * Convenience functions, in case only one needs to be overridden.
 * Also useful for setting a new default, e.g., invoking
 * NaClLogPreInitSetVerbosity with the maximum of the verbosity level
 * supplied from higher level code such as chrome's command line
 * flags, and the default value from the environment as returned by
 * NaClLogDefaultLogVerbosity().
 */
int NaClLogDefaultLogVerbosity(void);
struct Gio *NaClLogDefaultLogGio(void);

/*
 * Sets the log file to the named file.  Aborts program if the open
 * fails.  A GioFile object is associated with the file.
 *
 * The GioFile object is dynamically allocated, so caller is
 * responsible for obtaining it via NaClLogGetGio and freeing it as
 * appropriate, since otherwise a memory leak will occur.  This
 * includes closing the wrapped FILE *, if appropriate (e.g., not the
 * default of stderr).
 */
void NaClLogSetFile(char const *log_file);

void NaClLogModuleFini(void);

void NaClLogSetVerbosity(int verb);

int NaClLogGetVerbosity(void);

void NaClLogIncrVerbosity(void);

void NaClLogSetGio(struct Gio *out_stream);

struct Gio *NaClLogGetGio(void);

/*
 * Timestamps on log entries may be disabled, e.g., to make it easier to
 * write test that compare logging output.
 */

void NaClLogEnableTimestamp(void);

void NaClLogDisableTimestamp(void);

/*
 * Users of NaClLogV should add ATTRIBUTE_FORMAT_PRINTF(m,n) to their
 * function prototype, where m is the argument position of the format
 * string and n is the position of the first argument to be consumed
 * by the format specifier.
 */
void NaClLogV(int         detail_level,
              char const  *fmt,
              va_list     ap);

void NaClLog(int         detail_level,
             char const  *fmt,
             ...) ATTRIBUTE_FORMAT_PRINTF(2, 3);

#define LOG_INFO    (-1)
#define LOG_WARNING (-2)
#define LOG_ERROR   (-3)
#define LOG_FATAL   (-4)

/*
 * Low-level logging code that requires manual lock manipulation.
 * Should be used carefully!
 */
void NaClLogLock(void);
void NaClLogUnlock(void);

/*
 * Caller is responsible for grabbing the NaClLog mutex lock (via
 * NaClLogLock) before calling NaClLogV_mu or NaClLog_mu and for
 * releasing it (via NaClLogUnlock) afterward.
 *
 * Users of NaClLogV_mu should also use ATTRIBUTE_FORMAT_PRINTF as
 * above.
 *
 * Only the first call to NaClLog_mu or NaClLogV_mu after the
 * NaClLogLock will get a timestamp tag prepended.  No newline
 * detection is done, so a multi-line output must be (1) split into
 * multiple calls to NaClLog_mu or NaClLogV_mu so that newlines are
 * the last character in the format string to the NaClLog_mu or
 * NaClLogV_mu calls, and (2) are followed by NaClLogTagNext_mu() so
 * that the next output will know to generate a tag.
 */
void NaClLogV_mu(int         detail_level,
                 char const  *fmt,
                 va_list     ap);

void NaClLog_mu(int         detail_level,
                char const  *fmt,
                ...) ATTRIBUTE_FORMAT_PRINTF(2, 3);

void NaClLogTagNext_mu(void);

/*
 * The function at |fn| is invoked when NaClLog's family of logging
 * functions are invoked with LOG_FATAL.  This defaults to NaClAbort
 * (see nacl_exit.h), but may be overridden to do other things prior
 * to aborting.  NB: it is not a good idea to depend on too much
 * within |fn|, since the application state is likely
 * inconsistent/unstable.  It is probably a good idea to only use
 * low-level routines or system calls directly.
 */
void NaClLogSetAbortBehavior(void (*fn)(void));

/*
 * Run the current abort behavior hook.
 */
void NaClLogRunAbortBehavior(void);

EXTERN_C_END

#endif  /* NATIVE_CLIENT_SRC_TRUSTED_PLATFORM_NACL_LOG_H__ */
