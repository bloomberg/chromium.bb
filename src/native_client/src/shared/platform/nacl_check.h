/*
 * Copyright (c) 2012 The Native Client Authors. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

/*
 * NaCl service runtime, check macros.
 */

#ifndef NATIVE_CLIENT_SRC_SHARED_PLATFORM_NACL_CHECK_H_
#define NATIVE_CLIENT_SRC_SHARED_PLATFORM_NACL_CHECK_H_

#include "native_client/src/shared/platform/nacl_log.h"

EXTERN_C_BEGIN

/*
 * We cannot use variadic macros since not all preprocessors provide
 * them.  Instead, we require uses of the CHECK and DCHECK macro write
 * code in the following manner:
 *
 *   CHECK(a == b);
 *
 * or
 *
 *   VCHECK(a == b, ("a = %d, b = %d, c = %s\n", a, b, some_cstr));
 *
 * depending on whether a printf-like, more detailed message in
 * addition to the invariance failure should be printed.
 *
 * NB: BEWARE of printf arguments the evaluation of which have
 * side-effects.  Any such will cause the program to behave
 * differently depending on whether debug mode is enabled or not.
 */
#define CHECK(bool_expr) do {                                        \
    if (!(bool_expr)) {                                              \
      NaClLog(LOG_FATAL, "Fatal error in file %s, line %d: !(%s)\n", \
              __FILE__, __LINE__, #bool_expr);                       \
    }                                                                \
  } while (0)

#define DCHECK(bool_expr) do {                                       \
    if (nacl_check_debug_mode && !(bool_expr)) {                     \
      NaClLog(LOG_FATAL, "Fatal error in file %s, line %d: !(%s)\n", \
              __FILE__, __LINE__, #bool_expr);                       \
    }                                                                \
  } while (0)

#define VCHECK(bool_expr, fn_arg) do {                               \
    if   (!(bool_expr)) {                                            \
      NaClLog(LOG_ERROR, "Fatal error in file %s, line %d: !(%s)\n", \
              __FILE__, __LINE__, #bool_expr);                       \
      NaClCheckIntern fn_arg;                                        \
    }                                                                \
  } while (0)

#define DVCHECK(bool_expr, fn_arg) do {                              \
    if (nacl_check_debug_mode && !(bool_expr)) {                     \
      NaClLog(LOG_ERROR, "Fatal error in file %s, line %d: !(%s)\n", \
              __FILE__, __LINE__, #bool_expr);                       \
      NaClCheckIntern fn_arg;                                        \
    }                                                                \
  } while (0)

/*
 * By default, nacl_check_debug mode is 0 in opt builds and 1 in dbg
 * builds, so DCHECKs are only performed for dbg builds, though it's
 * possible to change this (viz, as directed by a command line
 * argument) by invoking NaClCheckSetDebugMode.  CHECKs are always
 * performed.
 */
extern void NaClCheckSetDebugMode(int mode);

/*
 * This is a private variable, needed for the macro.  Do not reference
 * directly.
 */
extern int nacl_check_debug_mode;

/*
 * This is a private function, used by the macros above.  Do not
 * reference directly.
 */
extern void NaClCheckIntern(const char *fmt, ...);

EXTERN_C_END

#endif
