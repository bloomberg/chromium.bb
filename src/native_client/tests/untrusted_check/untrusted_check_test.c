/*
 * Copyright (c) 2011 The Native Client Authors. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

#include "native_client/src/shared/platform/nacl_check.h"
#include "native_client/src/shared/platform/nacl_log.h"

/*
 * NB: platform_init is used in trusted code; unfortunately, the
 * untrusted version of the platform library is a subset of the
 * trusted platform library, and it is not currently possible to use
 * platform_init.h
 *
 * See http://code.google.com/p/nativeclient/issues/detail?id=1358 for
 * a sketch of what needs to be done to eliminate the need to invoke
 * NaClLogModuleInit explictly.
 *
 * #include "native_client/src/shared/platform/platform_init.h"
 */

int main(void) {
  /* NaClPlatformInit(); -- not available in untrusted platform lib */
  NaClLogModuleInit();
  NaClLog(LOG_INFO, "Log output\n");
  CHECK(0);
  return 0;
}
