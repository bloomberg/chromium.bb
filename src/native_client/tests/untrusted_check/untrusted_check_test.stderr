Log output
Fatal error in file untrusted_check_test.c
LOG_FATAL abort exit
